### DISCUSSION TITLE
What is the rendering api used by korge? opengl?

### DISCUSSION THREAD

**Discussion by ZXHHYJ** [korlibs/korge#1010] [Q&A] [answered]

**Comment by soywiz** [comment, marked as answer]
KorGE uses an abstraction called AG internally defined in the KorGW layer. That abstraction is used in korge and other layers. Right now AG is implemented in OpenGL supporting desktop OpenGL, mobile OpenGL|ES and WebGL, but can and will be implemented in other backends eventually.

Shaders are described in a Kotlin DSL that generates an AST, then right now it generates GLSL from that AST, but it is designed to be compatible with other backends as they appear and their matching shader languages. There are also plans to add a shader language parser that will be converted into that AST to be able to externally provide shaders without compiling, in a way that will also be compatible with all the implemented backends.
